### DISCUSSION TITLE
autocomplete not working with cmp, even though it works with every other language(on NixOS)?

### DISCUSSION THREAD

**Discussion by DrTeagle** [mrcjkb/rustaceanvim#523] [Q&A] [answered]
Howdy,
I am having an issue with autocomplete not showing up, maybe I am misunderstanding the capabilities, but everything else works, code actions works, and RustLsp logFile shows this:
```
2024-09-25T23:47:35.158808Z ERROR Flycheck failed to run the following command: CommandHandle { program: "/nix/store/a04bnqgj3mvg777a70jymwnxcgm7cdcv-rustup-1.27.1/bin/cargo", arguments: ["clippy", "--workspace", "--message-format=json-diagnostic-rendered-ansi", "--manifest-path", "/home/cdockter/Documents/CodingProjects/rustprojects/branches/Cargo.toml", "--keep-going", "--all-targets", "--all-features", "--no-deps"], current_dir: Some("/home/cdockter/Documents/CodingProjects/rustprojects/branches") }, error=Cargo watcher failed, the command produced no valid metadata (exit code: ExitStatus(unix_wait_status(256))):
error: 'cargo-clippy' is not installed for the custom toolchain '/nix/store/p024jn1sd4pn33lg0kaqxh7j2sp1iw7x-rustc-1.80.1'.
note: this is a custom toolchain, which cannot use `rustup component add`
help: if you built this toolchain from source, and used `rustup toolchain link`, then you may be able to build the component with `x.py`

2024-09-25T23:47:35.469745Z ERROR Flycheck failed to run the following command: CommandHandle { program: "/nix/store/a04bnqgj3mvg777a70jymwnxcgm7cdcv-rustup-1.27.1/bin/cargo", arguments: ["clippy", "--workspace", "--message-format=json-diagnostic-rendered-ansi", "--manifest-path", "/home/cdockter/Documents/CodingProjects/rustprojects/branches/Cargo.toml", "--keep-going", "--all-targets", "--all-features", "--no-deps"], current_dir: Some("/home/cdockter/Documents/CodingProjects/rustprojects/branches") }, error=Cargo watcher failed, the command produced no valid metadata (exit code: ExitStatus(unix_wait_status(256))):
error: 'cargo-clippy' is not installed for the custom toolchain '/nix/store/p024jn1sd4pn33lg0kaqxh7j2sp1iw7x-rustc-1.80.1'.
note: this is a custom toolchain, which cannot use `rustup component add`
help: if you built this toolchain from source, and used `rustup toolchain link`, then you may be able to build the component with `x.py`

```
however I do not think that these errors would cause autocomplete to stop working? i have autocomplete with every other language and have commented out my rustaceanvim config and it still seems to not work. I am not calling lspconfig at all to my knowledge. I am at a loss here. any pointers to fix this issue would be great. I have looked at similair discussions. and this is my capabilities block:
```
M.capabilities = require("cmp_nvim_lsp").default_capabilities(vim.lsp.protocol.make_client_capabilities())

M.capabilities.textDocument.completion.completionItem = {
	documentationFormat = { "markdown", "plaintext" },
	snippetSupport = true,
	preselectSupport = true,
	insertReplaceSupport = true,
	labelDetailsSupport = true,
	deprecatedSupport = true,
	commitCharactersSupport = true,
	tagSupport = { valueSet = { 1 } },
	resolveSupport = {
		properties = {
			"documentation",
			"detail",
			"additionalTextEdits",
		},
	},
}
```
when I call this and have no other config it still doesnt seem to work, i apologize profusely if there  is a trivial solution.

**Comment by mrcjkb** [comment]
hey 👋 

rustaceanvim auto-detects cmp_nvim_lsp and registers the capabilities, so you shouldn't need to do that manually. Although I doubt that's what's causing your issue. 

It's most likely related to your project or rust-analyzer build.
Do you have a minimal flake that you can reproduce it with? 
I also provide a derivation with a minimal config for Nix users. See the [troubleshooting section](https://github.com/mrcjkb/rustaceanvim?tab=readme-ov-file#stethoscope-troubleshooting).

**Comment by DrTeagle** [comment]
this is the flake I used for the dev shell:
```nix
{
  description = "rustacean flake";
  inputs = {

    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
    rustacean.url = "github:mrcjkb/rustaceanvim";
  };
  outputs = { self, nixpkgs, rustacean, ... }@inputs:
    let
      system = "x86_64-linux";
      pkgs = import nixpkgs { inherit system; };
    in
    {
      devShells.${system}.default = pkgs.mkShell
        {
          buildInputs = with pkgs;[
            clang
            libclang
            rocmPackages_6.llvm.clang-unwrapped
            libgcc
            glib
            gcc-unwrapped
            llvm_18
            pkg-config
            clang-tools
            rustacean.packages.${system}.codelldb
            rustc
            rustup
            rustfmt
            rust-analyzer
            cargo
          ];
          shellHook = ''
            rustup default stable
            rustup component add clippy
            export PATH="$HOME/.cargo/bin:$PATH"
            export PATH="$PATH:${rustacean.packages.${system}.codelldb}/bin"
            export PATH="${pkgs.rust-analyzer}/bin:$PATH"
            export PATH="$PATH:${pkgs.rustfmt}/bin"
          '';
        };
    };
}
```
 and I used the nix run on nightly and there was still no autocomplete

**Reply by mrcjkb** [reply]
That's not how I would set up a devShell for rust development. You shouldn't use rustup when using Nix. That sounds like a recipe for conflicts.
And there's no need to set up the `PATH` manually for packages in the devShell's `buildInputs`.

> and I used the nix run on nightly and there was still no autocomplete

For autocomplete you'll have to enable Neovim's built-in autocompletion or add a plugin for it to the package.

**Reply by DrTeagle** [reply]
I have this minimal.lua and I see no auto complete:
```lua
vim.env.LAZY_STDPATH = ".repro"
load(vim.fn.system("curl -s https://raw.githubusercontent.com/folke/lazy.nvim/main/bootstrap.lua"))()

require("lazy.minit").repro({
	spec = {
		{
			"hrsh7th/nvim-cmp",
			lazy = false,
		},
		{
			"hrsh7th/cmp-nvim-lsp",
			lazy = false,
		},
		{
			"mrcjkb/rustaceanvim",
			version = "^5",
			init = function()
				-- Configure rustaceanvim here
				vim.g.rustaceanvim = {}
			end,
			lazy = false,
		},
	},
})
```
also for the flake more tips on how to write it would be much appreciated, I am rather new to making dev shells with flakes.

**Reply by mrcjkb** [reply, marked as answer]
> I have this minimal.lua and I see no auto complete:

nvim-cmp doesn't work if you don't call its `setup` function.

> also for the flake more tips on how to write it would be much appreciated, I am rather new to making dev shells with flakes.

- Here's [an example of a rust project with a nix devShell](https://github.com/nvim-neorocks/rocks/blob/master/flake.nix).
- Here's [the template I use for creating a basic rust devShell](https://github.com/mrcjkb/flakes/blob/master/template/rust/flake.nix).

Instead of `pkgs` for your rust toolchain (cargo, rustc, ...), I recommend using `pkgs.rustPlatform`.

In general, there's not much I can do to help you when it comes to problems with autocompletion, because [rustaceanvim doesn't implement autocompletion](https://github.com/mrcjkb/rustaceanvim?tab=readme-ov-file#im-having-issues-with-autocompletion).
So the way to troubleshoot this is by bisecting your other plugins and troubleshooting your rust toolchain.

**Reply by DrTeagle** [reply]
figured it out cant reprouduce with minimal config. I called init for lazy and autocomplete works, marking this as closed

**Comment by jrray** [comment]
In case it helps anyone out there, I wasn't getting rust completions until I added this explicit dependency in my (lazy) configuration:

```lua
return {
    "mrcjkb/rustaceanvim",
    version = "^5", -- Recommended
    lazy = false,   -- This plugin is already lazy
    dependencies = {
        { "hrsh7th/cmp-nvim-lsp" },   -- <-- added this
    }
}
```
